### PR TITLE
Add StabilizerInverseSemigroup

### PR DESCRIPTION
**Pull request by ssiccha**
Adds an operation that returns the stabilizer of a point.
A method is installed for inverse semigroups of partial permutations.

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
Thanks for this too! I'll review it next week, I'm currently on holiday.

**Comment by ssiccha** [comment]
Just fixing the last things with respect to travis.

**Comment by ChrisJefferson** [comment]
I'm just dipping in as a non-semigroups person, but one comment I have is if that is really enough tests. I think you have the right idea with comparing to a brute force version, but could you do it for many more semigroups? Perhaps all (or many?) semigroups on a smallish number of points?

I do something similar in my packages, and it often knocks out special corner cases I didn't realise, like (for groups) on the symmetric group, or trivial group, or groups with interesting orbit structures, for some examples.

Of course, you might not want to always do that test if it is expensive, perhaps semigroups could have an expensive test directory for such things?

**Comment by james-d-mitchell** [comment]
I agree with @ChrisJefferson, I think it would be good to have some tests of the type Chris suggests. We have an `extreme` test folder where longer running tests can be placed. 

Also @wilfwilson I described in very rough terms how I would implement this to @ssiccha via slack, this is not written down anywhere. 

@ssiccha could you please revise this PR according to @wilfwilson's comments? Thanks!

File: gap/semigroups/semigrp.gi [hunk 1]
```diff
@@ -933,6 +933,93 @@ function(S, func)
...
+  fi;
+  orb := LambdaOrb(S);
+  scclookup := OrbSCCLookup(orb);
+  setswithpoint := Filtered([1 .. Size(orb)], i -> pnt in orb[i]);
+  scclookup := scclookup{setswithpoint};
+  tmp := List([1 .. Size(setswithpoint)],
+              i -> LambdaOrbMult(orb, scclookup[i], setswithpoint[i]));
+  sccreps := List(scclookup, i -> orb[OrbSCC(orb)[i][1]]);
+  fromsccrep := List(tmp, x -> x[1]);
+  tosccrep := List(tmp, x -> x[2]);
```
[the remark this answers is not in the document]
Comment (ssiccha) [reply gap/semigroups/semigrp.gi]: If always [this](https://github.com/gap-packages/Semigroups/blob/d4316ef2872003fc0101bb05c907fe4fb8c1dd21/gap/main/lambda-rho.gi#L139) is used in LambdaOrbMult then yes.

File: gap/semigroups/semigrp.gi [hunk 2]
```diff
@@ -933,6 +933,93 @@ function(S, func)
...
+  stabgens := [];
+  # Let x be a set containing pnt. We compute generators of the stabiliser
+  # of pnt in the Schutzenberger group of x. To this end, we use the set
+  # representing the SCC of x.
+  for i in [1 .. Length(setswithpoint)] do
+    to := tosccrep[i];
+    schutz := LambdaOrbSchutzGp(orb, scclookup[i]);
+    stab := Stabilizer(schutz, pnt ^ to);
+    # Convert the permutations in stab to PartialPerms
+    newstabgens := GeneratorsOfGroup(stab);
```
[the remark this answers is not in the document]
Comment (ssiccha) [reply gap/semigroups/semigrp.gi]: `GeneratorsOfGroup` now also returns the identity if the argument is a trivial group.

File: gap/semigroups/semigrp.gi [hunk 3]
```diff
@@ -933,6 +933,93 @@ function(S, func)
...
+  fromsccrep := List(tmp, x -> x[1]);
+  tosccrep := List(tmp, x -> x[2]);
+  stabgens := [];
+  # Let x be a set containing pnt. We compute generators of the stabiliser
+  # of pnt in the Schutzenberger group of x. To this end, we use the set
+  # representing the SCC of x.
+  for i in [1 .. Length(setswithpoint)] do
+    to := tosccrep[i];
+    schutz := LambdaOrbSchutzGp(orb, scclookup[i]);
+    stab := Stabilizer(schutz, pnt ^ to);
```
[the remark this answers is not in the document]
Comment (ssiccha) [reply gap/semigroups/semigrp.gi]: Mhm, should be possible. Let me see..

File: gap/semigroups/semigrp.gi [hunk 4]
```diff
@@ -933,6 +933,93 @@ function(S, func)
...
+    # Convert the permutations in stab to PartialPerms
+    newstabgens := GeneratorsOfGroup(stab);
+    dom := sccreps[i];
+    newstabgens := List(newstabgens, x -> AsPartialPerm(x, dom));
+    # If stab is trivial newstabgens will be empty. Add the identity pperm
+    if IsEmpty(newstabgens) then
+      newstabgens := [AsPartialPerm((), dom)];
+    fi;
+    # Conjugate them back into the Schutzenberger group of x
+    newstabgens := List(newstabgens, x -> x ^ (to ^ -1));
```
[the remark this answers is not in the document]
Comment (ssiccha) [reply gap/semigroups/semigrp.gi]: Done

File: gap/semigroups/semigrp.gi [hunk 5]
```diff
@@ -933,6 +933,93 @@ function(S, func)
...
+    Add(stabgens, newstabgens);
+  od;
+  stabgens := Concatenation(stabgens);
```
[the remark this answers is not in the document]
Comment (ssiccha) [reply gap/semigroups/semigrp.gi]: Done

File: gap/semigroups/semigrp.gi [hunk 6]
```diff
@@ -933,6 +933,93 @@ function(S, func)
...
+  # For two different sets in the same SCC take semigroup elements
+  # mapping one to the other via the SCC's representative.
+  # Then check whether pnt^to and PreImage(from, pnt) can be
+  # mapped to each other in the representatives Schutzenberger
+  # group.
+  for i in [1 .. Length(setswithpoint)] do
```
[the remark this answers is not in the document]
Comment (ssiccha) [reply gap/semigroups/semigrp.gi]: I just thought about it a little again. They may not necessary be equal bc `pi'` might not be `pi^-1`. But the result of `i -> j` is in the generated inverse semigroup iff the result of `j -> i` is.

I'll put into the comments why that's true.

**Comment by ssiccha** [comment]
I'll push the changes in a moment. What I want to do next is:
- snake_case all the local vars
- profile it again
- add some more tests

**Comment by ssiccha** [comment]
Hmm, can't get the new Semigroups version to run atm. I'll try again tomorrow.

**Comment by james-d-mitchell** [comment]
@ssiccha are you still having problems getting Semigroups to run?

**Comment by ssiccha** [comment]
@james-d-mitchell yes, testinstall seg faults always at the same line. I'll try to enable some debug info tomorrow and get back to you.

**Comment by james-d-mitchell** [comment]
Weird, are you sure you did `make clean && ./autogen.sh && ./configure && make -j4 && scripts/make-deps.py`?

**Comment by ssiccha** [comment]
@james-d-mitchell I didn't do `scripts/make-deps.py`. I suggest that semigroups checks whether that was run when its loaded.
The package crashes at line 1426 of `testinstall.tst`:
```
#T# Bug in AsInverseSemigroupCongruenceByKernelTrace - abnormal kernel
gap> S := InverseSemigroup([PartialPerm([1, 2], [1, 6]),
>                           PartialPerm([1, 2, 4, 5], [1, 4, 5, 3]),
>                           PartialPerm([1, 3, 4], [1, 2, 5]),
>                           PartialPerm([1, 2, 3, 5, 6], [5, 1, 6, 2, 4]),
>                           PartialPerm([1, 2, 3, 5, 6], [5, 4, 3, 1, 2])]);;
gap> pairs := [[PartialPerm([2, 5, 6], [1, 6, 5]),
>               PartialPerm([2, 5, 6], [5, 1, 6])],
>              [PartialPerm([5], [2]), PartialPerm([1, 5], [5, 4])]];;
gap> cong := SemigroupCongruence(S, pairs);;
gap> NrEquivalenceClasses(cong);
954
```

**Comment by james-d-mitchell** [comment]
@ssiccha I'm pretty sure there is no way to check from GAP that kernel extensions have been recompiled when you've updated GAP itself. The script `make-deps.py` is just a helper to recompile the dependencies of Semigroups for developers.

Just to clarify: are you still getting the seg faul after a fresh install? I can't reproduce this, and the travis tests are clear too.

**Comment by ssiccha** [comment]
I created a new worktree `Semigroups-clean` and checked it out to a copy of current master. It could have to do with the fact that its not in the `GAPROOT/pkg` directory. I'll test that now

**Comment by ssiccha** [comment]
@james-d-mitchell I made a fresh copy of semigroups. I still get the same error:
```
gap> Test("tst/testinstall.tst");
# line 1426 of 1776 (80%)Segmentation fault (core dumped)
```

Are there some debug info that I can send you?

**Comment by ssiccha** [comment]
@james-d-mitchell 
- Compiler: gcc 7.2.0
- OS: `cat /proc/version` yields `Linux version 4.13.3-1-ARCH (builduser@tobias) (gcc version 7.2.0 (GCC)) #1 SMP PREEMPT Thu Sep 21 20:33:16 CEST 2017`
- Semigroups: I'm using a fresh copy of [master](https://github.com/gap-packages/Semigroups/commit/1b81ae456f29df3e681f206ab839323e3d0a9e6a) from today
- libsemigroups: fresh copy of [master](11bcb5a25ef7a236f9b0b74e94a1027378b04729) also from today
- Report default options: actually, nothing was printed. This is my full run to reproduce the error. (semigroups version 3.0.X is just to make sure that I dont use something coming from `make bootstrap-pkg-full`)
```
 ┌───────┐   GAP 4.8.8-4943-g0c4601c of today
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-pc-linux-gnu-default64
 Configuration:  gmp 6.1.2, readline
 Loading the library and packages ...
Duplicate cookie io.c:IO_mkstemp
 Packages:   AClib 1.2, Alnuth 3.1.0, AtlasRep 1.5.1, AutPGrp 1.8, Browse 1.8.7, CRISP 1.4.4, 
             Cryst 4.1.13, CrystCat 1.1.6, CTblLib 1.2.2, FactInt 1.6.0, FGA 1.3.1, GAPDoc 1.6.1, 
             IO 4.5.0, IRREDSOL 1.4, LAGUNA 3.8.0, Polenta 1.3.8, Polycyclic 2.11, PrimGrp 3.3.0, 
             RadiRoot 2.7, ResClasses 4.7.1, smallgrp 1.2, Sophus 1.23, SpinSym 1.5, TomLib 1.2.6, 
             TransGrp 2.0.2, utils 0.49
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
gap> LoadPackage("semi");
────────────────────────────────────────────────────────────────────────────────────────────────────
Loading  orb 4.8.0 (Methods to enumerate orbits)
by Juergen Mueller (http://www.math.rwth-aachen.de/~Juergen.Mueller),
   Max Neunhöffer (http://www-groups.mcs.st-and.ac.uk/~neunhoef), and
   Felix Noeske (http://www.math.rwth-aachen.de/~Felix.Noeske).
Homepage: https://gap-packages.github.io/orb
────────────────────────────────────────────────────────────────────────────────────────────────────
────────────────────────────────────────────────────────────────────────────────────────────────────
Loading  genss 1.6.5 (Generic Schreier-Sims)
by Max Neunhöffer (http://www-groups.mcs.st-and.ac.uk/~neunhoef) and
   Felix Noeske (http://www.math.rwth-aachen.de/~Felix.Noeske).
Homepage: https://gap-packages.github.io/genss
────────────────────────────────────────────────────────────────────────────────────────────────────
────────────────────────────────────────────────────────────────────────────────────────────────────
Loading  GRAPE 4.7 (GRaph Algorithms using PErmutation groups)
by Leonard H. Soicher (http://www.maths.qmul.ac.uk/~leonard/).
Homepage: http://www.maths.qmul.ac.uk/~leonard/grape/
────────────────────────────────────────────────────────────────────────────────────────────────────
────────────────────────────────────────────────────────────────────────────────────────────────────
Loading  Digraphs 0.11.0 (Digraphs - Methods for digraphs)
by J. De Beule (http://homepages.vub.ac.be/~jdbeule/),
   J. Jonusas (http://www-groups.mcs.st-andrews.ac.uk/~julius/),
   J. D. Mitchell (http://goo.gl/ZtViV6),
   M. Torpey (http://www-groups.mcs.st-andrews.ac.uk/~mct25/), and
   W. A. Wilson (http://www-groups.mcs.st-andrews.ac.uk/~waw7/).
Homepage: https://gap-packages.github.io/Digraphs
────────────────────────────────────────────────────────────────────────────────────────────────────
-----------------------------------------------------------------------------
Loading  Semigroups 3.0.X
by J. D. Mitchell (http://www-groups.mcs.st-andrews.ac.uk/~jamesm/)
with contributions by:
     S. Burrell (http://sburrell.nfshost.com),
     M. Delgado (http://cmup.fc.up.pt/cmup/mdelgado/),
     J. East (http://goo.gl/MuiJu5),
     A. Egri-Nagy (http://www.egri-nagy.hu),
     N. Ham (https://n-ham.github.io),
     J. Jonusas (http://www-groups.mcs.st-andrews.ac.uk/~julius/),
     M. Pfeiffer (https://www.morphism.de/~markusp/),
     C. Russell,
     B. Steinberg (http://www.sci.ccny.cuny.edu/~benjamin/),
     J. Smith (http://math.sci.ccny.cuny.edu/people?name=Jhevon_Smith),
     M. Torpey (http://www-groups.mcs.st-and.ac.uk/~mct25/),
 and W. A. Wilson (http://wilf.me).
-----------------------------------------------------------------------------
true
gap> SEMIGROUPS.DefaultOptionsRec.report := true;
true
gap> Test("tst/testinstall.tst");
# line 1426 of 1793 (79%)Segmentation fault (core dumped)
```

**Comment by ssiccha** [comment]
Is there some kind of stacktrace I can get from somewhere?

**Comment by james-d-mitchell** [comment]
You could try running Semigroups in the debugger lldb or gdb.

**Comment by james-d-mitchell** [comment]
I'll investigate but can't right now

**Comment by james-d-mitchell** [comment]
Or if you can't do that, maybe just run the example that is failing with the reporting switched on, as indicated earlier.

**Comment by ssiccha** [comment]
No worries. I've never needed gdb or lldb. So now I'll have a look. Shouldn't be too hard. _fingers crossed_

**Comment by james-d-mitchell** [comment]
You could also try compiling Semigroups with the `./configure --enable-debug` which might end in an assertion failure which would then possibly lead to the issue.

**Comment by james-d-mitchell** [comment]
One more thing: this part of the code hasn't changed for several releases of libsemigroups, nor has the way we call it from Semigroups, is it possible that something else in your setup has changed?

**Comment by ssiccha** [comment]
Hmmm. I probably compiled hpcgap in that folder at _some_ point in time. But I did `make clean` and compiled GAP with no extra configuration before I ran it in the debugger. As you can see from the banner a few posts ago it _does not_ start hpcgap. Maybe hpcgap left some traces somewhere when I compiled it? This would then be another issue. How does gap / semigroups decide whether it runs in hpcgap mode? 

I'll try a fresh copy of gap then next.

**Comment by james-d-mitchell** [comment]
@ssiccha any progress?

**Comment by ssiccha** [comment]
@james-d-mitchell Weird. A fresh GAP installation also doesn't change anything. I still get the same seg fault. Doesn't semigroups use threads by default?

Also, what is this supposed to be in '#2'? `/build/gcc/src/gcc/libstdc++-v3/src/c++11/thread.cc:163`. There doesn't exist a directory /build on that computer and neither does it exist in the gap directory.
```
sergio ~/projects/gap-clean(gap-clean)* $ find . -name build
./hpcgap/extern/gc/build
sergio ~/projects/gap-clean(gap-clean)* $ ls ./hpcgap/extern/gc/build
s60v3
sergio ~/projects/gap-clean(gap-clean)* $ find . -name libstdc++-v3
sergio ~/projects/gap-clean(gap-clean)* $
```

**Comment by james-d-mitchell** [comment]
@ssiccha, yes Semigroups uses the STL `std::thread` objects, which ought to work just fine with the version of GCC you are using. 

I'm not sure what you're asking about #2, that's part of the implementation of `std::thread` on your computer, not part of libsemigroups, so I have no idea what should be in it. `/build/gcc/src/gcc/libstdc++-v3/src/c++11/thread.cc` is the file containing the implementation of `std::thread`, and must be on your computer, because there's a line number next to it, so it's getting read from somewhere. I'm somewhat convinced this is an issue with your computer, and not with Semigroups/libsemigroups itself.

Can you try running `make check` inside the `src/libsemigroups/` and see what happens?

**Comment by ssiccha** [comment]
K, I wil ask our admin where that `/build/` directory comes from. It's not in `/`.

`make check` returns `All checks passed`...

**Comment by ssiccha** [comment]
Hah, I think I worked on the first version of my PR on my laptop. That would explain why I didnt have this problem before. Unfortunately my laptop's SSD gave up today so I'll have to find some other machine to install semigroups on.

**Comment by ssiccha** [comment]
So, my SSD still is a bit finicky but it works again, at least atm. So maybe I'll be able to finish this up tomorrow.

**Comment by ssiccha** [comment]
Looks like my "proof" is incorrect. :thinking:

**Comment by ssiccha** [comment]
phew, it was just a "search & replace" error

**Comment by ssiccha** [comment]
Profiles look good. The most time is spent in trying to find a small generating set for the result.

**Comment by ssiccha** [comment]
Next: add more tests

**Comment by ssiccha** [comment]
@james-d-mitchell Travis reports that the outputs of `StabilizerInverseSemigroup` and `BruteForceStabilizerInverseSemigroup` differ. When I run the test on my laptop everything runs fine. Could you try the code on one of your machines? Do you have any idea what the reason could be?

**Comment by james-d-mitchell** [comment]
@ssiccha thanks for this, I'll test it out as soon as I get the chance.
